polygeist/include/mlir/Dialect/Polygeist/Analysis/DataFlowSolverWrapper.h

```diff
@@ -0,0 +1,64 @@
+//===- DataFlowSolverWrapper.h - Wrapper for a DataFlowSolver ----*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains utilities that allows dataflow analysis to declare and
+// load any dataflow analysis they might require.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_POLYGEIST_ANALYSIS_DATAFLOWSOLVERWRAPPER_H
+#define MLIR_DIALECT_POLYGEIST_ANALYSIS_DATAFLOWSOLVERWRAPPER_H
+
+#include "mlir/Analysis/AliasAnalysis.h"
+#include "mlir/Analysis/DataFlowFramework.h"
+
+namespace mlir {
+namespace polygeist {
+
+class DataFlowSolverWrapper;
+
+/// An interface that should be implemented by dataflow analysis in
+/// order to declare which other dataflow analyses they might require.
+template <class Derived> class RequiredDataFlowAnalyses {
+public:
+  /// Load dataflow analyses required by the \tparam Derived class in the \p
+  /// solver.
+  /// Note: derived classes must implement the static member function
+  /// 'loadRequiredAnalysesImpl'.
+  static void loadRequiredAnalyses(DataFlowSolverWrapper &solver) {
+    Derived::loadRequiredAnalysesImpl(solver);
+  }
+};
+
+/// Provides a way to load into a DataFlowSolver a dataflow analysis, along
+/// with any other dataflow analyses it might require.
+class DataFlowSolverWrapper : public DataFlowSolver {
+public:
+  DataFlowSolverWrapper(AliasAnalysis &aliasAnalysis)
```

Comment:
Do we want to allow data flow analyses that don't require AliasAnalysis?

Reply:
Not possible because *some* analysis might depend on a dataflow analysis that requires aliasAnalysis. 
As an example consider `UniformityAnalysis`. It needs the `ReachingDefinitionAnalysis` which needs to be constructed with an alias analysis object. The wrapper calls `loadRequiredAnalysesImpl` on `UniformityAnalysis` (a static member function) and pass a solver. The implementation in of `loadRequiredAnalysesImpl` must load the Reaching def. analysis and so it can only get the alias analysis from the solver.

I'd say for now "this is the way" (feeling like a Mandalorian today ahaha) ... but perhaps we can find a way to make the aliasAnalysis optional.

Reply:
ok, we can revisit when we want to use the `DataFlowSolverWrapper` with an analysis that doesn't require `AliasAnalysis`.